### ISSUE TITLE
 OpenTofu plans should show set members that will be retained post-apply

### ISSUE THREAD

**Issue by ktham** [opentofu/opentofu#2296]
### OpenTofu Version

```shell
1.6.0
```

### The problem in your OpenTofu project

OpenTofu defers rendering set comparisons until apply-time when there are new unknown elements adding to a resource attributes of `TypeSet`

For example, we might have configuration that initially looks like this:

```
resource "aws_security_group" "example" {
  # ...
}

resource "aws_db_instance" "example" {
  vpc_security_group_ids = [
    aws_security_group.example.id,
  ]
  # ...
}
```

And then we update the configuration to be this:
```
resource "aws_security_group" "example" {
  # ...
}

resource "aws_security_group" "additional" {
  # ...
}

resource "aws_db_instance" "example" {
  vpc_security_group_ids = [
    aws_security_group.example.id,
    aws_security_group.additional.id,
  ]
  # ...
}
```

Because `aws_security_group.additional.id` is not known until apply-time, and because `aws_db_instance.example.vpc_security_group_ids` is an attribute of `TypeSet`, we'd see OpenTofu Plan output that looks something like this:

```
      ~ vpc_security_group_ids                = [
          - "sg-46c08c0f",
        ] -> (known after apply)
```

This isn't a real or complete OpenTofu configuration, but is meant to highlight in brevity the UX issue we're facing.

Removal of security groups on an AWS resource like `aws_db_instance` could potentially be dangerous and it would give the OpenTofu practitioner more peace of mind if we can be assured from the OpenTofu Plan that the change above is either additive, or a no-op.

Here is desired OpenTofu Plan output that we'd like to see:
```
      ~ vpc_security_group_ids                = [
          "sg-46c08c0f",
          # (additional elements possible, but not known until after apply)
        ]
```

### Attempted Solutions

The workaround is a targeted apply of the resources that are unknown, in order to avoid the scenario of adding unknown elements to an existing set attribute, but that is not a feasible or scalable solution.

### Proposal

OpenTofu plans should show set members that are known to persist during the situation where new set members could be introduced (but is not yet known if they will be until apply-time).

This is important so that the practitioner can have assurance that certain set members are _not_ expected to be removed. This is particularly important when modifying configuration that impacts network connectivity.

However, OpenTofu today doesn't show anything at all, other than the TypeSet attribute will be assigned a new Set and the new Set will be "known after apply."

**Comment by apparentlymart** [comment]
Hi @ktham! Thanks for sharing this.

It is technically possible for OpenTofu to represent the situation you've described using [the type system library that the OpenTofu language is built on](https://github.com/zclconf/go-cty). The most accurate way to represent this situation in that library's terms would be the value produced by the following construction:

```go
cty.SetVal([]cty.Value{
    cty.StringVal("sg-46c08c0f"),
    cty.UnknownVal(cty.String),
})
```

I believe that the plan renderer already knows how to render that value as something like the following:

```
        vpc_security_group_ids = [
            "sg-46c08c0f",
            (known after apply),
        ]
```

or, when compared against your prior state value, like this:

```
      ~ vpc_security_group_ids = [
            "sg-46c08c0f",
          + (known after apply),
        ]
```

A set constructed in this way has an unknown [`length`](https://opentofu.org/docs/language/functions/length/) (because technically that second element could turn out to be also `"sg-46c08c0f"`, and so would coalesce with the known element) but can still enumerate the values that _are_ known, and report that it contains at least one unknown value.

Unfortunately I think the real problem here is in the `hashicorp/aws` provider, which AFAIK still uses the legacy Terraform plugin SDK to implement `aws_db_instance`, and so it's describing its plan in terms of the much-more-limited type system from Terraform v0.11 and earlier, where it was only possible to represent either an entirely-known or an entirely-unknown set.

Although I can't prove it with the provider in its current form, I expect that if this provider were updated to return the wire-protocol equivalent of the value I wrote out at the top then OpenTofu's plan renderer would already be able to produce output like my final example above. If that's true then unfortunately we can't really do anything about it from OpenTofu's side, since OpenTofu shouldn't make stronger promises than the provider made about what the final result will be. :confounded:

**Comment by ktham** [comment]
> Unfortunately I think the real problem here is in the hashicorp/aws provider, which AFAIK still uses the legacy Terraform plugin SDK to implement aws_db_instance, and so it's describing its plan in terms of the much-more-limited type system from Terraform v0.11 and earlier, where it was only possible to represent either an entirely-known or an entirely-unknown set.

Ah... I see... Thank you so much for your quick response and for your explanation, that makes sense. So I take it to mean that this can be fixed with this resource being implemented under [terraform plugin framework](https://github.com/hashicorp/terraform-plugin-framework)?

I was going to see if I might be able to take a stab at experimenting with it locally on my own to confirm your hypothesis, but taking a quick cursory glance at lines of code in [internal/service/rds/instance.go](https://github.com/hashicorp/terraform-provider-aws/blob/v5.81.0/internal/service/rds/instance.go), it looks like a pretty sizable resource so I likely won't attempt that anytime soon. I'll at the try to find time see if there is a Github issue in the hashicorp/aws provider about migrating `aws_db_instance` to the the latest plugin framework, and if not, then try to submit an issue

**Comment by apparentlymart** [comment]
Terraform Plugin Framework has its own type system that is compatible with but independent of `cty`, so I'm not 100% sure what is the equivalent to the `cty.SetVal(...)` call I showed in my earlier comment but I _think_ it's possible to express something equivalent to it using some combination of [`types.SetValue`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v1.13.0/types#SetValue), [`types.StringValue`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v1.13.0/types#StringValue), and [`types.StringUnknown`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v1.13.0/types#StringUnknown).

Indeed though, a rewrite of a resource type this complicated from legacy SDK to plugin framework is a significant amount of work and so not something I'd expect someone to do unless they were being paid to do it. :grinning: 

I did also peep in the repository to see if there was an existing issue for this and didn't find one, but perhaps I didn't look hard enough... there are _a lot_ of issues in there.
